api/v1beta1/conversion.go

```diff
@@ -16,14 +16,187 @@ limitations under the License.
+		Domain:               in.Spec.Domain,
+	}
+
+	out.Status = CloudStackClusterStatus{
+		FailureDomains: in.Status.FailureDomains,
+		Ready:          in.Status.Ready,
+	}
+	return nil
+}
+
+// getZones maps failure domains to zones
+func getZones(csCluster *v1beta2.CloudStackCluster) []Zone {
+	var zones []Zone
+	for _, failureDomain := range csCluster.Spec.FailureDomains {
+		zone := failureDomain.Zone
+		zones = append(zones, Zone{
+			Name: zone.Name,
+			ID:   zone.ID,
+			Network: Network{
+				Name: zone.Network.Name,
+				ID:   zone.Network.ID,
+				Type: zone.Network.Type,
+			},
+		})
+	}
+	return zones
+}
+
+// getFailureDomains maps v1beta1 zones to v1beta2 failure domains.
+func getFailureDomains(csCluster *CloudStackCluster) ([]v1beta2.CloudStackFailureDomainSpec, error) {
+	var failureDomains []v1beta2.CloudStackFailureDomainSpec
+	namespace := csCluster.Namespace
+	for _, zone := range csCluster.Spec.Zones {
+		name, err := GetDefaultFailureDomainName(namespace, csCluster.Name, zone.ID, zone.Name)
+		if err != nil {
+			return nil, err
+		}
+		failureDomains = append(failureDomains, v1beta2.CloudStackFailureDomainSpec{
+			Name: name,
+			Zone: v1beta2.CloudStackZoneSpec{
+				ID:   zone.ID,
+				Name: zone.Name,
+				Network: v1beta2.Network{
+					ID:   zone.Network.ID,
+					Name: zone.Network.Name,
+					Type: zone.Network.Type,
+				},
+			},
+			Domain:  csCluster.Spec.Domain,
+			Account: csCluster.Spec.Account,
+			ACSEndpoint: corev1.SecretReference{
+				Namespace: csCluster.ObjectMeta.Namespace,
+				Name:      DefaultEndpointCredential,
+			},
+		})
+	}
+	return failureDomains, nil
+}
+
+// GetDefaultFailureDomainName return zoneID as failuredomain name.
+// Default failure domain name is used when migrating an old cluster to a multiple-endpoints supported cluster, that
+// requires to convert each zone to a failure domain.
+// When upgrading cluster using eks-a, a secret named global will be created by eks-a, and it is used by following
+// method to get zoneID by calling cloudstack API.
+// When upgrading cluster using clusterctl directly, zoneID is fetched directly from kubernetes cluster in cloudstackzones.
+func GetDefaultFailureDomainName(namespace string, clusterName string, zoneID string, zoneName string) (string, error) {
+	if len(zoneID) > 0 {
+		return zoneID + "-" + clusterName, nil
+	}
+
+	// try fetch zoneID using zoneName through cloudstack client
+	zoneID, err := fetchZoneIDUsingCloudStack(namespace, zoneName)
+	if err == nil {
+		return zoneID + "-" + clusterName, nil
+	}
+
+	zoneID, err = fetchZoneIDUsingK8s(namespace, clusterName, zoneName)
+	if err != nil {
+		return "", nil
+	}
+	return zoneID + "-" + clusterName, nil
+}
+
+func fetchZoneIDUsingK8s(namespace string, clusterName string, zoneName string) (string, error) {
+	zones := &CloudStackZoneList{}
+	capiClusterLabel := map[string]string{clusterv1.ClusterLabelName: clusterName}
+	if err := v1beta2.K8sClient.List(
+		context.TODO(),
```

Comment:
what is this context?

Reply:
```
TODO returns a non-nil, empty Context. Code should use context.TODO when it's unclear which Context to use or it is not yet available (because the surrounding function has not yet been extended to accept a Context parameter).
```

Reply:
TODO returns a non-nil, empty Context. Code should use context.TODO when it's unclear which Context to use or it is not yet available (because the surrounding function has not yet been extended to accept a Context parameter).